base/loading.jl

```diff
@@ -1981,15 +1983,71 @@ function __require_prelocked(uuidkey::PkgId, env=nothing)
         # After successfully loading, notify downstream consumers
         run_package_callbacks(uuidkey)
     else
+        warn_if_already_loaded_different(uuidkey)
         newm = root_module(uuidkey)
     end
     return newm
 end
 
+const already_warned_path_change_pkgs = Set{PkgId}()
+# warns if the loaded version of a module is different to the one that locate_package wants to load
+function warn_if_already_loaded_different(uuidkey::PkgId)
+    uuidkey ∈ already_warned_path_change_pkgs && return
+    pkgorig = get(pkgorigins, uuidkey, nothing)
+    if pkgorig !== nothing && pkgorig.path !== nothing
+        # `locate_package()` can throw an error; if that happens just skip
+        local new_path
+        try
+            new_path = locate_package(uuidkey)
+        catch
+            return
+        end
+
+        # new_path can be `nothing` if `uuidkey` is not loadable in this environment
+        if new_path === nothing
+            return
+        end
+        if !samefile(fixup_stdlib_path(pkgorig.path), new_path)
+            # TODO: only check treehash if not same path, as it's expensive
+            loaded_treehash = 1 # TODO
+            new_treehash = 2 # TODO
```

Comment:
@fatteneder I wondered if you know of a cheap way to get these  treehashes for the loaded and new source paths?

Reply:
I am afraid you have to call `parse_cache_header` here to get an accurate result.